File: src/handlers/totalHolders.js
/**
 * Handler for the /total-holders endpoint
 */

import { executeGraphQLQuery } from '../utils/graphql.js';
import { json } from '../utils/response.js';

/**
 * Handler for GET /total-holders
 *
 * 1) Run a GraphQL query to count all holders.
 * 2) Return JSON { totalHolders: <number> }.
 * 
 * @param {Request} request - The original request
 * @param {FetchEvent} event - The fetch event
 * @returns {Promise<Response>} JSON response with totalHolders
 */
export async function totalHoldersHandler(request, event) {
  try {
    const holdersQuery = `
      query {
        allStates(
          filter: {
            key: { startsWith: "currency.balances:", notLike: "%:%:%" }
          }
        ) {
          totalCount
        }
      }
    `;
    
    const data = await executeGraphQLQuery(
      holdersQuery, 
      {}, 
      "Upstream GraphQL error on total-holders query"
    );
    
    const totalCountRaw = data?.data?.allStates?.totalCount;
    const totalHolders = totalCountRaw != null ? parseInt(totalCountRaw, 10) : 0;

    return json({ totalHolders });
  } catch (error) {
    // If error is already a Response (from executeGraphQLQuery), return it
    if (error instanceof Response) {
      return error;
    }
    // Otherwise, create a new error response
    return json({ error: error.message || "Internal error" }, { status: 500 });
  }
}

File: app/handlers/total_holders.py
"""
Handler for total holders endpoint
"""
import logging
from fastapi import Request

from app.utils.graphql import execute_graphql_query
from app.utils.response import json_response

logger = logging.getLogger(__name__)

async def total_holders_handler(request: Request):
    """
    Handler for GET /total-holders
    
    Returns the total number of holders across all tokens.
    """
    try:
        # GraphQL query to count all holders
        holders_query = """
            query {
                allStates(
                    filter: {
                        key: { startsWith: "currency.balances:", notLike: "%:%:%" }
                    }
                ) {
                    totalCount
                }
            }
        """
        
        # Execute query
        data = await execute_graphql_query(
            holders_query,
            {},
            "Upstream GraphQL error on total-holders query"
        )
        
        # Extract and parse total count
        total_count_raw = data.get('data', {}).get('allStates', {}).get('totalCount')
        total_holders = int(total_count_raw) if total_count_raw is not None else 0
        
        # Return response
        return json_response({"totalHolders": total_holders})
    
    except Exception as err:
        logger.error(f"Error in total_holders_handler: {err}", exc_info=True)
        return json_response(
            {"error": str(err) or "Internal error"},
            status_code=500
        )